### DISCUSSION TITLE
Interactive plotting with Shiny Apps

### DISCUSSION THREAD

**Discussion by berlianaputri2** [teamalgoritma/community#782] [Q&A] [answered]
Why error when trying to publish the shinyapps?
![image](https://github.com/teamalgoritma/community/assets/148056240/9c078db9-af50-46eb-952a-7379ae18f2cb)

**Comment by saskia-dwi-ulfah** [comment]
Dear Bu @berlianaputri2 ,
Terima kasih atas pertanyaannya.

Kalau berdasarkan error yang muncul, kemungkinan Bu Berliana membuka rmd-nya bukan dari R project capstone-nya. Bisa dicoba untuk mengakses semua file capstone-nya melalui R-project dan mengulangi lagi tahapan untuk running di local-nya. Jika error masih sama, boleh kirimkan file .zip capstone-nya ke email mentor@algorit.ma untuk kami lihat lebih lanjut.

Semoga jawaban ini dapat membantu.

Best regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
Sekarang sudah tidak error. Tapi pada bagian dataset, sebelumnya data sudah keluar, namun ketika di-close dan di-run kembali dataset tidak muncul meskipun kode sama. Kenapa ya? 
![image](https://github.com/teamalgoritma/community/assets/148056240/71ef7f8c-4fa6-484e-9ee1-c61411f13ac4)

**Reply by saskia-dwi-ulfah** [reply]
Boleh tolong dilampirkan kode yang bagian untuk menampilkan tablenya seperti apa, Bu?

**Reply by berlianaputri2** [reply]
library(shiny)

  shinyServer(function(input, output) {
    output$stroke <- DT::renderDataTable(
      stroke_filter)
      #options = list(scrollX = T, scrollY = T)
  
 
  # --- plot 1: bar
  output$plot1_bar <- renderPlotly({
    # Plot 1 Bar: Stroke Count based on Age Distribution and Gender
    
    # data transformation
    stroke_filter <- stroke %>% filter (stroke == 1)
    stroke_filter <- stroke_filter %>%
      mutate(Age_status = case_when(
        between(age, 1, 24) ~ "< 25 years",
        between(age, 25, 34) ~ "25-34 years",
        between(age, 35, 44) ~ "35-44 years",
        between(age, 45, 54) ~ "45-54 years",
        between(age, 55, 64) ~ "55-64 years",
        between(age, 65, 84) ~ ">65 years"
      ))
    
    age_gender <- stroke_filter %>%
      group_by(Age_status, gender) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot statis
    plot1 <- ggplot(stroke_filter, aes(x = Age_status, fill = gender)) +
      geom_bar(position = "dodge") +
      labs(title = "Stroke Count based on Age Distribution and Gender", x = "Age Group", y = "Number of People") +
      theme_minimal()+
      theme(legend.position = "none")
    plot1 <- ggplotly(plot1, tootltip = "x")
    plot1
    
    # membuat plot interaktif
    ggplotly(plot1, tootltip = "x")
  })
  
  
  output$plot2_bar <- renderPlotly({
    # Plot 2 Bar: Stroke Count based on Heart Disease, Hypertension, and Glucose Level
    
    # data transformation
    stroke_filter <- stroke_filter %>%
      mutate(Glucose_level_status = case_when(
        avg_glucose_level < 90 ~ "Low Blood Glucose Level",
        between(avg_glucose_level, 90, 130) ~ "Normal Blood Glucose Level",
        TRUE ~ "High Blood Glucose Level"
      ))
    
    health_contributors <- stroke_filter %>%
      group_by(hypertension, heart_disease, Glucose_level_status) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot bar - ggplot: bar plot
    plot2<- ggplot(health_contributors, aes(x = hypertension, y = stroke_count, fill = heart_disease, color = Glucose_level_status)) +
      geom_bar(stat = "identity", position = "dodge") +
      labs(title = "Stroke Count based on Heart Disease, Hypertension, and Glucose Level",
           x = "Heart Disease",
           y = "Stroke Count") +
      scale_fill_discrete(name = "Heart Disease") +
      scale_color_discrete(name = "Glucose Level") +
      theme_minimal()
    
    # plot interaktif: ggplotly
    ggplotly(plot2, tooltip = c("x", "y"))
  })
  
  
  
  output$plot3_bar <- renderPlotly({
    # Plot 3 Bar: Stroke Count based on Residence and Work Type**
    
    # data transformation
    stress_contributors <- stroke_filter %>%
      group_by(Residence_type, work_type) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot statis - ggplot
    plot3 <- ggplot(stress_contributors, aes(x = Residence_type, y = stroke_count, fill = work_type)) +
      geom_bar(stat = "identity", position = "dodge") +
      labs(title = "Stroke Count based on Residence and Work Type",
           x = "Residence Type",
           y = "Stroke Count") +
      theme_minimal()+
      theme(legend.position = "none")
    
    # plot interaktif: ggplotly
    ggplotly(plot3, tootltip = "x")
  })
  
  
  output$plot4_bar <- renderPlotly({
    # Plot 4 Bar: Stroke Count based on Body Mass Index (BMI) and Smoking Habit**
    
    # data transformation
    stroke_filter <- stroke_filter %>%
      mutate(BMI_status = case_when(
        bmi < 18.5 ~ "Underweight",
        between(bmi, 18.5, 24.9) ~ "Healthy range",
        between(bmi, 25, 29.9) ~ "Overweight",
        between(bmi, 30, 39.9) ~ "Obesity",
        TRUE ~ "Severe Obesity"
      ))
    
    habit_contributors <- stroke_filter %>%
      group_by(smoking_status, BMI_status) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot bar - ggplot
    plot4 <- ggplot(stroke_filter, aes(x = smoking_status, fill = BMI_status)) +
      geom_bar(position = "dodge") +
      labs(title = "Stroke Count based on BMI and Smoking Habit", x = "Smoking Habit", y = "Number of People") +
      theme_minimal()+
      theme(legend.position = "none")
    
    # plot interaktif: ggplotly
    ggplotly(plot4, tootltip = "x")
  })
})
  
Kode dimasukkan ke dalam server.

**Reply by jafarhf1** [reply]
Dear Ibu @berlianaputri2 ,

Mengenai kode di bagian penampilan dataset sepertinya ada kesalahan penulisan yaitu kurang tanda `{}` dan juga penggunaan fungsi `datatable`. Kodenya harusnya dituliskan seperti ini 
```
output$stroke <- DT::renderDataTable({
   datatable(stroke_filter, options = list(scrollX = T, scrollY = T))
})
```
Semoga membantu,

Best Regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
Tampilan data set masih tidak keluar.

![image](https://github.com/teamalgoritma/community/assets/148056240/bfd8f7f5-4778-470b-9343-c9ef211c6838)
![Uploading image.png…]()

**Reply by jafarhf1** [reply]
Dear @berlianaputri2,

Apakah Ibu mendefinisikan dataframe `stroke_filter` nya di dalam renderPlotly(), jika iya maka memang tidak bisa terbaca Bu, mungkin bisa coba didefinisikannya di file `global.R`. Semoga membantu

Best Regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
apakah di global.R didefinisikan seperti ini?  datatable <- stroke_filter

**Reply by jafarhf1** [reply]
Dear Ibu @berlianaputri2,

Tidak perlu ya Bu, maksudnya adalah mendefenisikan variabel dataframe nya seperti misalnya `stroke_filter <- read.csv()` supaya variablel tersebut bisa dibaca oleh renderDataTable nya.

Best Regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
Sudah dilakukan juga di global.R, tetapi dataset tidak muncul

**Reply by jafarhf1** [reply]
Dear Ibu @berlianaputri2 ,

Terima kasih atas konfirmasinya, boleh tolong dilampirkan Bu kode di bagian server dan ui nya?

Best regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
global. R = 

library(flexdashboard)
library(dplyr)
library(ggplot2)
library(scales)
library(glue)
library(plotly)
library(lubridate)
options(scipen = 100)

stroke <- read.csv("healthcare-dataset-stroke-data.csv")

stroke$id <- as.character(stroke$id)
stroke$hypertension <- as.factor(stroke$hypertension)
stroke$heart_disease <- as.factor (stroke$heart_disease)
stroke$stroke <- as.factor (stroke$stroke)
stroke$bmi <- as.numeric(stroke$bmi)
stroke <- na.omit(stroke)

stroke_filter <- stroke %>% filter (stroke == 1)
stroke_filter <- stroke_filter %>%
  mutate(Age_status = case_when(
    between(age, 1, 24) ~ "< 25 years",
    between(age, 25, 34) ~ "25-34 years",
    between(age, 35, 44) ~ "35-44 years",
    between(age, 45, 54) ~ "45-54 years",
    between(age, 55, 64) ~ "55-64 years",
    between(age, 65, 84) ~ ">65 years"
  ))

age_gender <- stroke_filter %>%
  group_by(Age_status, gender) %>%
  count(stroke) %>%
  rename(stroke_count = n)

stroke_filter <- stroke_filter %>%
  mutate(Glucose_level_status = case_when(
    avg_glucose_level < 90 ~ "Low Blood Glucose Level",
    between(avg_glucose_level, 90, 130) ~ "Normal Blood Glucose Level",
    TRUE ~ "High Blood Glucose Level"
  ))

health_contributors <- stroke_filter %>%
  group_by(hypertension, heart_disease, Glucose_level_status) %>%
  count(stroke) %>%
  rename(stroke_count = n)

stress_contributors <- stroke_filter %>%
  group_by(Residence_type, work_type) %>%
  count(stroke) %>%
  rename(stroke_count = n)

stroke_filter <- stroke_filter %>%
  mutate(BMI_status = case_when(
    bmi < 18.5 ~ "Underweight",
    between(bmi, 18.5, 24.9) ~ "Healthy range",
    between(bmi, 25, 29.9) ~ "Overweight",
    between(bmi, 30, 39.9) ~ "Obesity",
    TRUE ~ "Severe Obesity"
  ))

habit_contributors <- stroke_filter %>%
  group_by(smoking_status, BMI_status) %>%
  count(stroke) %>%
  rename(stroke_count = n)

ui. R = 
library(shiny)
library(shinydashboard)

dashboardPage(
  skin = "blue",
  
  dashboardHeader(title = "Study Trends of Stroke"),
  
  dashboardSidebar(
    sidebarMenu(
      menuItem("Overview", tabName = "page1", icon = icon("dashboard")),
      menuItem("Stroke Risk Analysis", tabName = "page2", icon = icon("heartbeat")), # Change "stroke" to a valid icon
      menuItem("Dataset", tabName = "page3", icon = icon("database"))
    )
  ),
  
  dashboardBody(
    tabItems(
      tabItem(
        tabName = "page1",
        fluidRow(
          infoBox(
            width = 6,
            title = "Total People with Stroke",
            color = "maroon",
            value = nrow(stroke_filter)  
          )
        ),
        fluidRow(
          box(
            width = 12,
            plotlyOutput("plot1_bar")  
          )
        )
      ),
      
      tabItem(
        tabName = "page2",
        fluidRow(
          box(
            width = 6,
            plotlyOutput("plot2_bar")  
          ),
          box(
            width = 6,
            plotlyOutput("plot3_bar") 
          ),
          box(
            width = 6,
            plotlyOutput("plot4_bar")  
          )
        )
      ),
      
      tabItem(
        tabName = "page3",
        fluidRow(
          box(
            width = 12,
            dataTableOutput("datatable")  
          )
        )
      )
    )
  )
)

server.R = 

library(shiny)

  shinyServer(function(input, output) {
    output$stroke <- DT::renderDataTable({
      datatable(stroke_filter, options = list(scrollX = T, scrollY = T))
    })
  
 
  # --- plot 1: bar
  output$plot1_bar <- renderPlotly({
    # Plot 1 Bar: Stroke Count based on Age Distribution and Gender
    
    # data transformation
    stroke_filter <- stroke %>% filter (stroke == 1)
    stroke_filter <- stroke_filter %>%
      mutate(Age_status = case_when(
        between(age, 1, 24) ~ "< 25 years",
        between(age, 25, 34) ~ "25-34 years",
        between(age, 35, 44) ~ "35-44 years",
        between(age, 45, 54) ~ "45-54 years",
        between(age, 55, 64) ~ "55-64 years",
        between(age, 65, 84) ~ ">65 years"
      ))
    
    age_gender <- stroke_filter %>%
      group_by(Age_status, gender) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot statis
    plot1 <- ggplot(stroke_filter, aes(x = Age_status, fill = gender)) +
      geom_bar(position = "dodge") +
      labs(title = "Stroke Count based on Age Distribution and Gender", x = "Age Group", y = "Number of People") +
      theme_minimal()+
      theme(legend.position = "none")
    plot1 <- ggplotly(plot1, tootltip = "x")
    plot1
    
    # membuat plot interaktif
    ggplotly(plot1, tootltip = "x")
  })
  
  
  output$plot2_bar <- renderPlotly({
    # Plot 2 Bar: Stroke Count based on Heart Disease, Hypertension, and Glucose Level
    
    # data transformation
    stroke_filter <- stroke_filter %>%
      mutate(Glucose_level_status = case_when(
        avg_glucose_level < 90 ~ "Low Blood Glucose Level",
        between(avg_glucose_level, 90, 130) ~ "Normal Blood Glucose Level",
        TRUE ~ "High Blood Glucose Level"
      ))
    
    health_contributors <- stroke_filter %>%
      group_by(hypertension, heart_disease, Glucose_level_status) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot bar - ggplot: bar plot
    plot2<- ggplot(health_contributors, aes(x = hypertension, y = stroke_count, fill = heart_disease, color = Glucose_level_status)) +
      geom_bar(stat = "identity", position = "dodge") +
      labs(title = "Stroke Count based on Heart Disease, Hypertension, and Glucose Level",
           x = "Heart Disease",
           y = "Stroke Count") +
      scale_fill_discrete(name = "Heart Disease") +
      scale_color_discrete(name = "Glucose Level") +
      theme_minimal()
    
    # plot interaktif: ggplotly
    ggplotly(plot2, tooltip = c("x", "y"))
  })
  
  
  
  output$plot3_bar <- renderPlotly({
    # Plot 3 Bar: Stroke Count based on Residence and Work Type**
    
    # data transformation
    stress_contributors <- stroke_filter %>%
      group_by(Residence_type, work_type) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot statis - ggplot
    plot3 <- ggplot(stress_contributors, aes(x = Residence_type, y = stroke_count, fill = work_type)) +
      geom_bar(stat = "identity", position = "dodge") +
      labs(title = "Stroke Count based on Residence and Work Type",
           x = "Residence Type",
           y = "Stroke Count") +
      theme_minimal()+
      theme(legend.position = "none")
    
    # plot interaktif: ggplotly
    ggplotly(plot3, tootltip = "x")
  })
  
  
  output$plot4_bar <- renderPlotly({
    # Plot 4 Bar: Stroke Count based on Body Mass Index (BMI) and Smoking Habit**
    
    # data transformation
    stroke_filter <- stroke_filter %>%
      mutate(BMI_status = case_when(
        bmi < 18.5 ~ "Underweight",
        between(bmi, 18.5, 24.9) ~ "Healthy range",
        between(bmi, 25, 29.9) ~ "Overweight",
        between(bmi, 30, 39.9) ~ "Obesity",
        TRUE ~ "Severe Obesity"
      ))
    
    habit_contributors <- stroke_filter %>%
      group_by(smoking_status, BMI_status) %>%
      count(stroke) %>%
      rename(stroke_count = n)
    
    # membuat plot bar - ggplot
    plot4 <- ggplot(stroke_filter, aes(x = smoking_status, fill = BMI_status)) +
      geom_bar(position = "dodge") +
      labs(title = "Stroke Count based on BMI and Smoking Habit", x = "Smoking Habit", y = "Number of People") +
      theme_minimal()+
      theme(legend.position = "none")
    
    # plot interaktif: ggplotly
    ggplotly(plot4, tootltip = "x")
  })
})

**Reply by jafarhf1** [reply]
Dear Ibu @berlianaputri2,

Terima kasih telah melampirkan kodenya. Berdasarkan hal tersebut, sepertinya permasalahannya terdapat pada tidak cocoknya antara id yang dituliskan di `ui.R` dengan `server.R` nya. Pada file `server.R`, Ibu menuliskannya adalah `output$stroke <- ....` yang mana berarti id nya adalah "stroke" sedangkan pada file `ui.R` Id yang dimasukkan di dalam fungsi `dataTableOutput` adalah "datatable" sehingga tidak cocok. Oleh karena itu, silahkan dicocokkan terlebih dahulu.

Demikian yang dapat kami sampaikan, semoga bisa membantu

Best regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
Outputnya malah seperti ini kak. Apa boleh file saya kirimkan ke email untuk diperiksa lebih lanjut? karena kemarin dataset sudah berhasil ditampilkan, tapi ketika di-run ulang malah tidak muncul.
![image](https://github.com/teamalgoritma/community/assets/148056240/e9cbdee7-79e2-4759-871f-767e4e8a83c5)

**Reply by jafarhf1** [reply]
Dear @berlianaputri2,

Mengenai hal tersebut bisa coba untuk mengimport library DT terlebih dahulu atau bisa juga langsung ditambahkan `DT::datatable()` di dalam fungsi `renderDataTable` nya ya. Apabila masih terdapat kendala boleh silahkan dikirimkan filenya ya Bu Terima kasih.

Best regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
Oke, saya kirimkan melalui email ya kak. Terima kasih.

**Reply by jafarhf1** [reply, marked as answer]
Dear @berlianaputri2 

Berdasarkan pengamatan kami terhadap file yang Ibu berikan, sepertinya memang permasalahan muncul disebabkan Ibu menggunakan fungsi-fungsi dari library DT tetapi tidak diimport terlebih dahulu library tersebut. Oleh karena itu, apabila tidak melakukan import, silahkan tambahkan DT:: di depan fungsi yang akan digunakan. Terdapat dua bagian kode yang membuat tampilan data tidak muncul di shiny dan berikut adalah perbaikan yang perlu dilakukan:

1. Bagian pertama terletak di file server.R 
```
​output$datatable <- DT::renderDataTable({
      DT::datatable(stroke_filter)
```
2. Bagian yang kedual terletak di file ui.R
```
box(
    width = 12,
    DT::dataTableOutput("datatable")  # Include server logic for the dataset
)
```

Silahkan untuk diubah pada bagian tersebut.

​Demikian yang dapat kami sampaikan, semoga membantu


Best Regards,
Team Algoritma

**Reply by berlianaputri2** [reply]
Oke, terima kasih!
